firrtl/src/main/scala/firrtl/ir/IR.scala

```diff
@@ -539,8 +539,49 @@ object GroundType {
 }
 abstract class AggregateType extends Type
 
-case class ProbeType(underlying: Type) extends Type with UseSerializer
-case class RWProbeType(underlying: Type) extends Type with UseSerializer
+// TODO switch back to case class when we make this private
+class ProbeType(val underlying: Type, val color: Option[String])
+    extends Type
+    with UseSerializer
+    with Product
+    with Serializable {
+  def this(underlying: Type) = this(underlying, None)
+  def copy(underlying: Type = underlying): ProbeType = ProbeType(underlying, color)
+  def canEqual(that:   Any):               Boolean = that.isInstanceOf[ProbeType]
+  def productArity: Int = 2
+  def productElement(n: Int): Any = n match {
+    case 0 => this.underlying
+    case 1 => this.color
+    case x => throw new IndexOutOfBoundsException(x.toString)
+  }
+}
+object ProbeType extends scala.runtime.AbstractFunction1[Type, ProbeType] {
+  def unapply(a:        ProbeType): Option[Type] = Some((a.underlying))
+  def apply(underlying: Type): ProbeType = new ProbeType(underlying)
+  def apply(underlying: Type, color: Option[String]): ProbeType = new ProbeType(underlying, color)
+}
+
+// TODO switch back to case class when we make this private
+class RWProbeType(val underlying: Type, val color: Option[String])
+    extends Type
+    with UseSerializer
+    with Product
+    with Serializable {
+  def this(underlying: Type) = this(underlying, None)
+  def copy(underlying: Type = underlying): RWProbeType = RWProbeType(underlying, color)
+  def canEqual(that:   Any):               Boolean = that.isInstanceOf[RWProbeType]
+  def productArity: Int = 2
+  def productElement(n: Int): Any = n match {
+    case 0 => this.underlying
+    case 1 => this.color
+    case x => throw new IndexOutOfBoundsException(x.toString)
+  }
+}
+object RWProbeType extends scala.runtime.AbstractFunction1[Type, RWProbeType] {
+  def unapply(a:        RWProbeType): Option[Type] = Some(a.underlying)
+  def apply(underlying: Type): RWProbeType = new RWProbeType(underlying)
+  def apply(underlying: Type, color: Option[String]): RWProbeType = new RWProbeType(underlying, color)
+}
```

Comment:
```suggestion
final case class ProbeType(val underlying: Type, val color: Option[String] = None) extends Type with UseSerializer
final class RWProbeType(val underlying: Type, val color: Option[String] = None) extends Type with UseSerializer
```
Since this is not going to be backported, let's not fight with binary compatibility.

Reply:
Note that the `final` makes it possible to make them `@data` later which is handy.

Reply:
Will switch to this.